### DISCUSSION TITLE
Custom data types

### DISCUSSION THREAD

**Discussion by roreshko** [sql-formatter-org/sql-formatter#594] [Q&A] [answered]
I apply the formatter for our custom SQL-like language, and it does a good job.

But in our dialect, we have a version data type with values like 3.14.5, and the latest versions of formatter throw an error:
```
Parse error at token: 1 at line 1 column 176
Unexpected NUMBER token: {"type":"NUMBER","raw":"1","text":"1","start":175}. Instead, I was expecting to see one of the following:...

```
Is there any possibility of extending the formatter with additional data types?

**Comment by nene** [comment]
You can define a custom dialect. See: https://github.com/sql-formatter-org/sql-formatter/blob/master/docs/dialect.md#custom-dialect-configuration-experimental

There's no real way of defining a custom data-type per se, but you can make the formatter treat it as a string, by defining a regex to match it:

```js
tokenizerOptions: {
  stringTypes: [
    "''", // the string types of the SQL dialect you're extending
    { regex: "\d+.\d+.\d+" }, // regex to match version numbers
  ],
}
```

**Reply by roreshko** [reply]
As you proposed, I tried the following approach:

` var d = sqlFormatter.snowflake;
   d.tokenizerOptions.stringTypes = ['$$', "''-qq-bs", { regex: '\\d+.\\d+.\\d+' }];
   var formatted = sqlFormatter.formatDialect('source_client_version>=4.36.1', { dialect: d });`
   
But the result is still the same:

`Parse error at token: 1 at line 1 column 29
Unexpected NUMBER token: {"type":"NUMBER","raw":"1","text":"1","start":28}. Instead, I was expecting to see one of the following:
`
As I see it parses 4.36.1 as NUMBER [still](https://user-images.githubusercontent.com/128801039/234582515-b0698063-5529-4874-963a-5076c1753fc1.png)

**Reply by nene** [reply, marked as answer]
Hmm... yeah, it happens that the tokenizer tries to match NUMBER token before STRING, so it ends up matching `4.36` as a NUMBER.

But I think you can instead use `identTypes` (to treat it as an identifier), which gets matched before number.
